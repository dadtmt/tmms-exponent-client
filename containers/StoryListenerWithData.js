import R from 'ramda'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import { pageEditorId } from '../config'
import StoryListener from '../components/StoryListener'

const GET_PAGE_EDITOR_QUERY = gql`
query GetPageEditor($pageEditorId: ID!) {
  getPageEditor(id: $pageEditorId) {
    id
    crossroads(first: 1, orderBy: {field:createdAt, direction:DESC}){
      edges{
        node{
          id
          isReady
          text
          choices{
            edges{
              node{
                id
                text
                made
              }
            }
          }
          testDices{
            edges{
              node{
                id
                details
                nbDices
                nbSides
                made
                master
                modifier
                text
                result
              }
            }
          }
        }
      }
    }
  }
}
`
const withData = graphql(
  GET_PAGE_EDITOR_QUERY,
  {
    options: {
      variables: { pageEditorId }
    }
  }
)

export const isChoiceMade = R.pipe(
  R.pathOr([], ['getPageEditor', 'crossroads', 'edges']),
  R.head,
  R.pathOr([], ['node', 'choices', 'edges']),
  R.map(R.path(['node', 'made'])),
  R.reduce(
    R.or,
    false
  )
)

export const isReady = R.pipe(
  R.pathOr([], ['getPageEditor', 'crossroads', 'edges']),
  R.head,
  R.pathOr(false, ['node', 'isReady'])
)

export const isActive = R.converge(
  R.and,
  [isReady, R.complement(isChoiceMade)]
)

const MAKE_CHOICE_MUTATION = gql`
mutation UpdateChoice($madeChoice: UpdateChoiceInput!) {
  updateChoice(input: $madeChoice){
    changedChoice {
      id
      made
      text
    }
  }
}
`

const withMakeChoice = graphql(
  MAKE_CHOICE_MUTATION,
  {
    props: ({ mutate, ownProps: { data } }) => ({
      makeChoice: choice => {
        if(isActive(data)) {
          mutate({
            variables: {
              madeChoice: {
                id: choice.id,
                made: true
              }
            }
          })
        }
      }
    })
  }
)

const RESOLVE_TEST_MUTATION = gql`
mutation ResolveTest($resolvedTest: UpdateTestDiceInput!) {
  updateTestDice(input: $resolvedTest){
    changedEdge {
      node {
        details
        id
        made
        master
        modifier
        nbDices
        nbSides
        result
      }
    }
  }
}
`

const withResolveTest = graphql(
  RESOLVE_TEST_MUTATION,
  {
    props: ({ mutate, ownProps: { data } }) => ({
      resolveTest: testDice => {
        if(isActive(data)) {
          mutate({
            variables: {
              resolvedTest: {
                id: testDice.id,
                made: true
              }
            }
          })
        }
      }
    })
  }
)

export default R.compose(
  withData,
  withMakeChoice,
  withResolveTest
)(StoryListener)
