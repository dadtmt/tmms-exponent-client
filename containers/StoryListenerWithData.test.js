import React from 'react'
import renderer from 'react-test-renderer'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo'

import StoryListenerWithData, {
  isActive,
  isChoiceMade,
  isReady
} from './StoryListenerWithData'

const client = new ApolloClient({
  networkInterface: {
    subscribe: jest.fn()
  }
})

describe('isChoiceMade', () => {
  it('return true if choice is made', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [
            {
              node: {
                choices: { edges: [
                  {
                    node: {
                      made: true
                    }
                  },
                  {
                    node: {
                      made: false
                    }
                  }
                ] }
              }
            }
          ]
        }
      }
    }
    expect(isChoiceMade(data)).toBeTruthy()
  })
  it('return false if choice is not made', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [
            {
              node: {
                choices: { edges: [
                  {
                    node: {
                      made: false
                    }
                  },
                  {
                    node: {
                      made: false
                    }
                  }
                ] }
              }
            }
          ]
        }
      }
    }
    expect(isChoiceMade(data)).toBeFalsy()
  })
  it('return false if no current crossroads', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: []
        }
      }
    }
    expect(isChoiceMade(data)).toBeFalsy()
  })
  it('return false if loading', () => {
    const data = {
      loading: true
    }
    expect(isChoiceMade(data)).toBeFalsy()
  })
})

describe('isReady', () => {
  it('return true if is ready', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [{
            node: {
              isReady: true
            }
          }]
        }
      }
    }
    expect(isReady(data)).toBeTruthy()
  })
  it('return false if is not ready', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [{
            node: {
              isReady: false
            }
          }]
        }
      }
    }
    expect(isReady(data)).toBeFalsy()
  })
})

describe('isActive', () => {
  it('return true if is ready', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [{
            node: {
              isReady: true
            }
          }]
        }
      }
    }
    expect(isActive(data)).toBeTruthy()
  })
  it('return false if is not ready', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [{
            node: {
              isReady: false
            }
          }]
        }
      }
    }
    expect(isActive(data)).toBeFalsy()
  })
  it('return false if ready and choice is made', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [{
            node: {
              choices: { edges: [
                {
                  node: {
                    made: true
                  }
                },
                {
                  node: {
                    made: false
                  }
                }
              ] },
              isReady: true
            }
          }]
        }
      }
    }
    expect(isActive(data)).toBeFalsy()
  })
  it('return false if is not ready and choice is not made', () => {
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [{
            node: {
              choices: { edges: [
                {
                  node: {
                    made: false
                  }
                },
                {
                  node: {
                    made: false
                  }
                }
              ] },
              isReady: false
            }
          }]
        }
      }
    }
    expect(isActive(data)).toBeFalsy()
  })
})

describe('StoryListenerWithData', () => {
  it('renders correctly', () => {
    expect(
      renderer.create(
        <ApolloProvider client={client}>
          <StoryListenerWithData />
        </ApolloProvider>
      ).toJSON()
    ).toMatchSnapshot()
  })
})
