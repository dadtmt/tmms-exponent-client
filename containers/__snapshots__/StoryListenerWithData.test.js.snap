exports[`StoryListenerWithData renders correctly 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Loading...
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Attends ton tour...
  </Text>
  <View />
</View>
`;
